feat(gemma3): Add BiGemma3 and ColGemma3 models with Matryoshka embeddings

adithya-s-k opened the pull request:
## Summary

This PR adds support for **BiGemma3** and **ColGemma3** models based on the Gemma3-4B-IT backbone, enabling multilingual multimodal document retrieval.

### Key Features

- **BiGemma3**: Single-vector dense retrieval model with Matryoshka representation learning
  - Supports flexible embedding dimensions (768, 1536, 2560) at inference time
  - Efficient document retrieval with configurable accuracy/speed trade-offs
  
- **ColGemma3**: Multi-vector late interaction model using ColBERT-style architecture
  - Fine-grained token-level matching with MaxSim scoring
  - 128-dimensional per-token embeddings

### Changes

- Added `BiGemma3` model in `colpali_engine/models/gemma3/bigemma3/`
  - `modeling_bigemma.py`: Model implementation with Matryoshka support
  - `processing_bigemma.py`: Processor for images and text
  
- Added `ColGemma3` model in `colpali_engine/models/gemma3/colgemma3/`
  - `modeling_colgemma.py`: Multi-vector model implementation
  - `processing_colgemma.py`: Processor with MaxSim scoring

- Added comprehensive tests in `tests/models/gemma3/`
  - Unit tests for model loading
  - Integration tests for forward pass
  - Retrieval tests with visual documents

### API Design

**BiGemma3** allows choosing embedding dimension at inference time:

```python
from colpali_engine.models import BiGemma3, BiGemmaProcessor3

model = BiGemma3.from_pretrained(
    "Cognitive-Lab/NetraEmbed",
    torch_dtype=torch.bfloat16,
    device_map="cuda",
)

# Choose dimension at inference time
embeddings = model(**inputs, embedding_dim=1536)
```

**ColGemma3** uses standard multi-vector late interaction:

```python
from colpali_engine.models import ColGemma3, ColGemmaProcessor3

model = ColGemma3.from_pretrained(
    "Cognitive-Lab/ColNetraEmbed",
    torch_dtype=torch.bfloat16,
    device_map="cuda",
)

embeddings = model(**inputs)  # (batch, num_patches, 128)
```

### Models

- [NetraEmbed (BiGemma3)](https://huggingface.co/Cognitive-Lab/NetraEmbed)
- [ColNetraEmbed (ColGemma3)](https://huggingface.co/Cognitive-Lab/ColNetraEmbed)

### Related Work

- Paper: [M3DR: Towards Universal Multilingual Multimodal Document Retrieval](https://arxiv.org/abs/2512.03514)
- Blog: [Introducing NetraEmbed](https://www.cognitivelab.in/blog/introducing-netraembed)
- Benchmark: [NayanaIR-Bench](https://huggingface.co/collections/Cognitive-Lab/nayanair-bench)

adithya-s-k commented:
@athrael-soju , i have pushed the code for the interpretability do check it out

ManuelFay reviewed (changes requested):
LGTM but can you please ruff the code and fix the tests so CI pass ?

adithya-s-k commented:
Hey i have just pushed two model 
These are not gated and should be easy to test they are just the base gemma model and for the col model , base model + the projection layers
- https://huggingface.co/Nayana-cognitivelab/bigemma
- https://huggingface.co/Nayana-cognitivelab/colgemma

also the final checkpoints can be used
- https://huggingface.co/Cognitive-Lab/NetraEmbed
- https://huggingface.co/Cognitive-Lab/ColNetraEmbed

adithya-s-k commented:
Hi @ManuelFay , I have made all the requested changes.
Updated the PR to include the ungated base models for both BiGemma3 and ColGemma3, fixed the model references in tests, and added the interpretability tests.

I have locally verified that the tests now run correctly with these models and everything looks good on my side.

ManuelFay commented:
Awesome, I'll merge my linting PR and then merge yours ! Thanks a ton !

ManuelFay commented:
can you rebase on main (will make the ruff CI happy) ? then we will merge !

ManuelFay reviewed (changes requested):
Only 1 ruff error remaining. Then we can merge, rest looks nice !
Thanks again !

adithya-s-k commented on tests/interpretability/test_interpretability_workflow.py:
```diff
@@ -0,0 +1,422 @@
+"""
+
+import math
+from typing import List, Tuple
```
Have made the changes

adithya-s-k commented:
@ManuelFay have run it locally and tested, it should pass all the checks now

ManuelFay commented:
Thank you for the contributions ! Don't hesitate to submit your model results on the MTEB visual retrieval leaderboard !
